python/ray/data/_internal/execution/operators/map_operator.py

```diff
@@ -0,0 +1,125 @@
+from typing import List, Iterator, Any, Dict, Callable, Optional
+
+import ray
+from ray.data.block import Block, BlockMetadata
+from ray.data._internal.stats import StatsDict
+from ray.data._internal.compute import (
+    ComputeStrategy,
+    TaskPoolStrategy,
+    ActorPoolStrategy,
+)
+from ray.data._internal.execution.interfaces import (
+    RefBundle,
+    PhysicalOperator,
+)
+from ray.data._internal.execution.operators.map_operator_tasks_impl import (
+    MapOperatorTasksImpl,
+)
+from ray.data._internal.execution.operators.map_operator_actors_impl import (
+    MapOperatorActorsImpl,
+)
+
+
+class MapOperator(PhysicalOperator):
+    """A streaming operator that maps input bundles 1:1 to output bundles.
+
+    This operator implements the distributed map operation, supporting both task
+    and actor compute strategies.
+    """
+
+    def __init__(
+        self,
+        transform_fn: Callable[[Iterator[Block]], Iterator[Block]],
+        input_op: PhysicalOperator,
+        name: str = "Map",
+        compute_strategy: Optional[ComputeStrategy] = None,
+        min_rows_per_batch: Optional[int] = None,
+        ray_remote_args: Optional[Dict[str, Any]] = None,
+    ):
+        """Create a MapOperator.
+
+        Args:
+            transform_fn: The function to apply to each ref bundle input.
+            input_op: Operator generating input data for this op.
+            name: The name of this operator.
+            compute_strategy: Customize the compute strategy for this op.
+            min_rows_per_batch: The number of rows to gather per batch passed to the
+                transform_fn, or None to use the block size. Setting the batch size is
+                important for the performance of GPU-accelerated transform functions.
+                The actual rows passed may be less if the dataset is small.
+            ray_remote_args: Customize the ray remote args for this op's tasks.
+        """
+        self._transform_fn = transform_fn
+        self._strategy = compute_strategy or TaskPoolStrategy()
+        self._remote_args = (ray_remote_args or {}).copy()
+        self._output_metadata: List[BlockMetadata] = []
+        self._min_rows_per_batch = min_rows_per_batch
+        if isinstance(self._strategy, TaskPoolStrategy):
+            self._execution_state = MapOperatorTasksImpl(self)
```

Comment:
Another abstraction hierarchy smell: we have a circular dependency between the operator abstraction and its execution state, with each depending on each other.